### PR TITLE
Fix output path verification for onnx-mlir and update the lit tests to not pollute the source tree

### PR DESCRIPTION
**Pull request by sstamenova**
Some lit tests leave artifacts in the source tree. Update the tests, so that all artifacts are placed in the binary tree.

This revealed a bug in the output path verification (checking whether the specified output path is valid), so the verification is updated as well. It now works correctly regardless of the type of slashes used as well as regardless of the depth of the paths.

Signed-off-by: Stella Stamenova <stilis@microsoft.com>

### CODE REVIEW

File: test/mlir/driver/check_passthrough_options.mlir [hunk 1]
```diff
@@ -1,13 +1,13 @@
-// RUN: onnx-mlir -Xopt --data-sections -v %s 2>&1 | FileCheck --check-prefix=OPT %s
-// RUN: onnx-mlir -Xllc --data-sections -v %s 2>&1 | FileCheck --check-prefix=LLC %s
-// RUN: onnx-mlir -mllvm --data-sections -v %s 2>&1 | FileCheck --check-prefix=LLVM %s
+// RUN: onnx-mlir -Xopt --data-sections -v %s -o %T/check_passthrough_options 2>&1 | FileCheck --check-prefix=OPT %s
```
Comment (caoimhinuibrian) [inline test/mlir/driver/check_passthrough_options.mlir:1]: The lit manual says %T is deprecated?

Comment (sstamenova) [reply test/mlir/driver/check_passthrough_options.mlir]: It does say so, but it is still used in llvm. I updated to the new recommendation anyway to make sure the tests aren't impacted in the future.

**Review by etiotto — APPROVED** [review]
LGTM

**Comment by etiotto** [comment]
@sstamenova yes I can run them on a Mac. I will take a look at the failure there.

**Comment by etiotto** [comment]
This is the error I get when we use a full path for `-o`:
```
%  /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/Debug/bin/onnx-mlir -Xopt --data-sections -v /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/test/mlir/driver/check_passthrough_options.mlir -o /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/test/mlir/driver/Output/check_passthrough_options.mlir.tmp/check_passthrough_options
libc++abi: terminating with uncaught exception of type std::__1::regex_error: The complexity of an attempted match against a regular expression exceeded a pre-set level.
zsh: abort      /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/Debug/bin/onnx-mlir   -
```

If I change the argument to `-o` to a shorter relative path the command completes successfully:
```
%  /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/Debug/bin/onnx-mlir -Xopt --data-sections -v /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/test/mlir/driver/check_passthrough_options.mlir -o ./check_passthrough_options.mlir.tmp/check_passthrough_options 
[/Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/test/mlir/driver/Output/]/Users/etiotto@ca.ibm.com/Source/onnx/llvm-project/build/bin/opt: opt -O0 --mtriple=x86_64-apple-darwin21.3.0 --data-sections -o ./check_passthrough_options.mlir.tmp/check_passthrough_options.bc ./check_passthrough_options.mlir.tmp/check_passthrough_options.unoptimized.bc
[/Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/test/mlir/driver/Output/]/Users/etiotto@ca.ibm.com/Source/onnx/llvm-project/build/bin/llc: llc -O0 --mtriple=x86_64-apple-darwin21.3.0 -filetype=obj -relocation-model=pic -o ./check_passthrough_options.mlir.tmp/check_passthrough_options.o ./check_passthrough_options.mlir.tmp/check_passthrough_options.bc
[/Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/test/mlir/driver/Output/]/usr/bin/clang++: clang++ ./check_passthrough_options.mlir.tmp/check_passthrough_options.o -o ./check_passthrough_options.mlir.tmp/check_passthrough_options.so -shared -fPIC -L/Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/Debug/lib -lcruntime
```

**Comment by etiotto** [comment]
Seems OK if the path is not too 'complex':
```
% /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/Debug/bin/onnx-mlir -Xopt --data-sections /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/test/mlir/driver/check_passthrough_options.mlir -o  /tmp/a/b/c/d/g/check_passthrough_options.mlir.tmp; echo $?
0
```

```
% /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/Debug/bin/onnx-mlir -Xopt --data-sections /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/test/mlir/driver/check_passthrough_options.mlir -o  /tmp/a/b/c/d/g/this@is.more.complex/check_passthrough_options.mlir.tmp; echo $?
0
```

But this fails (too many '/' seems to confuse `libc++abi` !!!!):
```
% /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/Debug/bin/onnx-mlir -Xopt --data-sections /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/test/mlir/driver/check_passthrough_options.mlir -o /tmp/a/b/c/d/e/f/g/h/i/j/k/l/m/n/o/p/q/r/s/t/u/v/w/x/y/z  ; echo $? 
libc++abi: terminating with uncaught exception of type std::__1::regex_error: The complexity of an attempted match against a regular expression exceeded a pre-set level.
```

**Comment by etiotto** [comment]
Not sure how to solve the problem, but  with `/tmp/a/b/c/d/e/f/g/h/i/j` it works, with one more subdirectory it fails: 
```
% /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/Debug/bin/onnx-mlir -Xopt --data-sections /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/test/mlir/driver/check_passthrough_options.mlir -o /tmp/a/b/c/d/e/f/g/h/i/j  ; echo $?
0
```
```
% /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/Debug/bin/onnx-mlir -Xopt --data-sections /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/test/mlir/driver/check_passthrough_options.mlir -o /tmp/a/b/c/d/e/f/g/h/i/j/k  ; echo $?
libc++abi: terminating with uncaught exception of type std::__1::regex_error: The complexity of an attempted match against a regular expression exceeded a pre-set level.
```

**Comment by sstamenova** [comment]
> @sstamenova yes I can run them on a Mac. I will take a look at the failure there.

Thanks @etiotto! I changed the tests to not add a new temporary directory and to directly use the automatically generated temporary file path. Hopefully that uncomplicates the path enough for mac.

**Pull request merged**